test/test_functionalization.py

```diff
@@ -0,0 +1,215 @@
+        input_clone2 = inpt.clone()
+        input_functional = torch._to_functional_tensor(input_clone2)
+
+        # Compare outputs (and mutated inputs), with and without functionalization.
+        out_ref = func(inpt)
+
+        torch._enable_functionalization()
+        try:
+            out_functional = func(input_functional)
+        finally:
+            torch._disable_functionalization()
+
+        # We need to sync the input tensors first, in case there are any queued mutations left.
+        torch._sync(input_functional)
+        torch._sync(out_functional)
+        self.assertEqual(out_ref, torch._from_functional_tensor(out_functional))
+        self.assertEqual(inpt, torch._from_functional_tensor(input_functional))  # input mutations should still occur
+
+    def test_simple(self):
+        def f(x):
+            # simple test: 1 view op, 1 inplace op
+            tmp = torch.ones(4, 2)
+            y = x.view(4, 2)
+            y.add_(tmp)
+            z = x * x
+            return y
+        self.assert_functionalization(f, torch.ones(4, 2))
+        logs = self.get_logs(f, torch.ones(4, 2))
+        self.assertExpectedInline('\n'.join(logs), """\
+$1 = torch._ops.aten.view($0, [4, 2])
+$2 = torch._ops.aten.add($1, tensor([[1., 1.],
+        [1., 1.],
+        [1., 1.],
+        [1., 1.]]))
+$3 = torch._ops.aten.view($2, [4, 2])
+$4 = torch._ops.aten.mul($3, $3)""")
+
+    def test_diagonal(self):
+        def f(x):
+            # test: view ops that take a subset of the original tensor (select/diagonal)
+            tmp = torch.ones(2)
+            y = x.diagonal()
+            y.add_(tmp)
+            z = x * x
+            return z
+        self.assert_functionalization(f, torch.ones(2, 2))
+        logs = self.get_logs(f, torch.ones(2, 2))
+        self.assertExpectedInline('\n'.join(logs), """\
+$1 = torch._ops.aten.diagonal($0)
+$2 = torch._ops.aten.add($1, tensor([1., 1.]))
+$3 = torch._ops.aten.diagonal_scatter($0, $2)
+$4 = torch._ops.aten.mul($3, $3)""")
+
+    def test_diagonal_mutated_input(self):
+        def f(x):
+            # simple test: there are pending updates afterwards, which the test syncs manually
+            tmp = torch.ones(2)
+            y = x.diagonal()
+            y.add_(tmp)
+            return x
+        x = torch.ones(2, 2)
+        self.assert_functionalization(f, x)
+
+    def test_copy__functionalized(self):
+        def f(x):
+            # simple test: functionalizing copy_() should call its out-of-place variant
+            tmp = torch.zeros(2, 2)
+            # NOTE: LoggingTensor isn't a mode, which means that the diagonal call
+            # will not be logged. This is fine for testing.
+            tmp_slice = tmp.diagonal()
+            tmp_slice.copy_(x)
+            return x
+        self.assert_functionalization(f, torch.ones(2))
+        logs = self.get_logs(f, torch.ones(2))
+        self.assertExpectedInline('\n'.join(logs), """$1 = torch._ops.aten.expand($0, [2])""")
```

Comment:
I added this test for `copy_()`, but the logging tensor output looks wrong (we should be logging `aten.to` before the expand call).

I'm hoping to look at this tomorrow, but I'm wondering if it's maybe a LoggingTensor interaction problem, instead of a bug in the previous PR. I already manually tested the conditional functionalization on diag_embed (from https://github.com/pytorch/functorch/pull/235), and that seems to work - and correctly call `to` and `expand_as`.

Reply:
This seems fine to me. LoggingTensor is logging at the `__torch_dispatch__` level. What this means is that it'll only log non-CompositeImplicitAutograd operations. When `Tensor::to` is a no-op, like here, it will not emit any CompositeImplicitAutograd operations; instead it just returns the original tensor as-is. A similar thing is happening with `expand_as`: we see `expand` in the Log because expand_as is CompositeImplicitAutograd but `expand` isn't.

Reply:
Good catch - completely forgot that if a composite kernel no-ops, LoggingTensor will not see anything.